File: app/src/main/java/dev/bee/kanjianki/MainActivitySettingsAutomationAutoSync.java
package dev.bee.kanjianki;

import android.view.View;
import android.widget.Toast;

import dev.bee.kanjianki.core.AutoSyncSettingsTogglePolicy;
import dev.bee.kanjianki.core.DateTextPolicy;
import dev.bee.kanjianki.core.SettingsTextCopy;
import dev.bee.kanjianki.sync.AutoSyncScheduler;

final class MainActivitySettingsAutomationAutoSync {
    private final MainActivitySettings activity;

    MainActivitySettingsAutomationAutoSync(MainActivitySettings activity) {
        this.activity = activity;
    }

    View autoSyncSettingsPanel() {
        return MainActivitySettingsAutomationAutoSyncCompose.autoSyncSettingsPanelView(
                activity,
                autoSyncSettingsPanelModel()
        );
    }

    SettingsAutoSyncPanelModel autoSyncSettingsPanelModel() {
        dev.bee.kanjianki.data.LocalStore.AutoSyncSettings auto = activity.store.autoSyncSettings();
        String lastSuccess = auto.lastSuccessAt > 0L ? DateTextPolicy.shortDateTime(auto.lastSuccessAt) : "";
        String lastAttempt = auto.lastAttemptAt > 0L && auto.lastAttemptAt != auto.lastSuccessAt
                ? DateTextPolicy.shortDateTime(auto.lastAttemptAt)
                : "";
        String nextRun = auto.nextRunAt > 0L ? DateTextPolicy.shortDateTime(auto.nextRunAt) : "";
        String actionLabel = null;
        boolean primaryAction = false;
        SettingsAutoSyncAction action = null;
        if (auto.configured) {
            if (auto.enabled) {
                actionLabel = SettingsTextCopy.turnOffDailySyncLabel();
                action = this::disableAutoSync;
            } else {
                actionLabel = SettingsTextCopy.turnOnDailySyncLabel();
                primaryAction = true;
                action = this::enableAutoSync;
            }
        }
        return new SettingsAutoSyncPanelModel(
                SettingsTextCopy.dailyAnkiSyncTitle(),
                SettingsTextCopy.autoSyncStatus(auto.configured, auto.enabled, auto.displayTime()),
                auto.enabled ? activity.TEAL : activity.MUTED,
                SettingsTextCopy.autoSyncDetail(auto.configured, auto.enabled, lastSuccess, lastAttempt, nextRun),
                actionLabel,
                primaryAction,
                action
        );
    }

    private void enableAutoSync() {
        AutoSyncSettingsTogglePolicy.ToggleResult result = AutoSyncSettingsTogglePolicy.enable();
        activity.store.setAutoSyncEnabled(result.enabled());
        AutoSyncScheduler.schedule(activity);
        Toast.makeText(activity, result.message(), Toast.LENGTH_SHORT).show();
        activity.renderSettings();
    }

    private void disableAutoSync() {
        AutoSyncSettingsTogglePolicy.ToggleResult result = AutoSyncSettingsTogglePolicy.disable();
        activity.store.setAutoSyncEnabled(result.enabled());
        AutoSyncScheduler.cancel(activity);
        Toast.makeText(activity, result.message(), Toast.LENGTH_SHORT).show();
        activity.renderSettings();
    }

}

File: app/src/main/kotlin/dev/bee/kanjianki/MainActivitySettingsAutomationAutoSync.kt
package dev.bee.kanjianki

import android.view.View
import android.widget.Toast
import dev.bee.kanjianki.core.AutoSyncSettingsTogglePolicy
import dev.bee.kanjianki.core.DateTextPolicy
import dev.bee.kanjianki.core.SettingsTextCopy
import dev.bee.kanjianki.data.LocalStoreBase
import dev.bee.kanjianki.sync.AutoSyncScheduler

internal class MainActivitySettingsAutomationAutoSync(private val activity: MainActivitySettings) {
    fun autoSyncSettingsPanel(): View {
        return autoSyncSettingsPanelView(activity, autoSyncSettingsPanelModel())
    }

    fun autoSyncSettingsPanelModel(): SettingsAutoSyncPanelModel {
        val auto = activity.store.autoSyncSettings()
        val action = autoSyncAction(auto)
        return SettingsAutoSyncPanelModel(
            title = SettingsTextCopy.dailyAnkiSyncTitle(),
            status = SettingsTextCopy.autoSyncStatus(auto.configured, auto.enabled, auto.displayTime()),
            statusColor = if (auto.enabled) MainActivityUiSupport.TEAL else MainActivityUiSupport.MUTED,
            detail = SettingsTextCopy.autoSyncDetail(
                auto.configured,
                auto.enabled,
                shortDateTime(auto.lastSuccessAt),
                lastAttemptText(auto),
                shortDateTime(auto.nextRunAt)
            ),
            actionLabel = action?.label,
            primaryAction = action?.primary ?: false,
            onAction = action?.callback
        )
    }

    private fun enableAutoSync() {
        val result = AutoSyncSettingsTogglePolicy.enable()
        activity.store.setAutoSyncEnabled(result.enabled())
        AutoSyncScheduler.schedule(activity)
        Toast.makeText(activity, result.message(), Toast.LENGTH_SHORT).show()
        activity.renderSettings()
    }

    private fun disableAutoSync() {
        val result = AutoSyncSettingsTogglePolicy.disable()
        activity.store.setAutoSyncEnabled(result.enabled())
        AutoSyncScheduler.cancel(activity)
        Toast.makeText(activity, result.message(), Toast.LENGTH_SHORT).show()
        activity.renderSettings()
    }

    private fun autoSyncAction(auto: LocalStoreBase.AutoSyncSettings): AutoSyncActionModel? {
        if (!auto.configured) {
            return null
        }
        return if (auto.enabled) {
            AutoSyncActionModel(
                label = SettingsTextCopy.turnOffDailySyncLabel(),
                primary = false,
                callback = SettingsAutoSyncAction { disableAutoSync() }
            )
        } else {
            AutoSyncActionModel(
                label = SettingsTextCopy.turnOnDailySyncLabel(),
                primary = true,
                callback = SettingsAutoSyncAction { enableAutoSync() }
            )
        }
    }

    private data class AutoSyncActionModel(
        val label: String,
        val primary: Boolean,
        val callback: SettingsAutoSyncAction
    )

    private companion object {
        fun shortDateTime(value: Long): String {
            return if (value > 0L) DateTextPolicy.shortDateTime(value) else ""
        }

        fun lastAttemptText(auto: LocalStoreBase.AutoSyncSettings): String {
            return if (auto.lastAttemptAt > 0L && auto.lastAttemptAt != auto.lastSuccessAt) {
                DateTextPolicy.shortDateTime(auto.lastAttemptAt)
            } else {
                ""
            }
        }
    }
}
